marshmallow/marshalling.py

```diff
@@ -250,57 +250,55 @@ def deserialize(
             return ret
         partial_is_collection = is_collection(partial)
         ret = dict_class()
-        for attr_name, field_obj in iteritems(fields_dict):
-            if field_obj.dump_only:
-                continue
-            field_name = attr_name
-            if field_obj.data_key:
-                field_name = field_obj.data_key
-            try:
+        # Check data is a dict
+        try:
+            data.get('dummy', missing)
+        except AttributeError:
+            errors = self.get_errors(index=index)
+            msg = 'Invalid input type.'
```

Comment:
This is fine for now until #852 is addressed. Might as well remove `'type'` in `Field.default_error_messages` while you're at it.